rust/src/storage/file/rename.rs

```diff
@@ -1,61 +1,81 @@
 use crate::StorageError;
-use std::ffi::CString;
-
-#[cfg(target_os = "linux")]
-const RENAME_NOREPLACE: libc::c_uint = 1;
-
-#[cfg(target_os = "linux")]
-extern "C" {
-    fn renameat2(
-        olddirfd: libc::c_int,
-        oldpath: *const libc::c_char,
-        newdirfd: libc::c_int,
-        newpath: *const libc::c_char,
-        flags: libc::c_uint,
-    ) -> libc::c_int;
-}
-
-pub fn rename(from: &str, to: &str) -> Result<(), StorageError> {
-    let ret = unsafe {
-        let from = to_c_string(from)?;
-        let to = to_c_string(to)?;
-        platform_specific_rename(from.as_ptr(), to.as_ptr())
-    };
 
-    if ret != 0 {
-        std::fs::remove_file(from).unwrap_or(());
+#[cfg(windows)]
+mod imp {
+    use super::*;
 
-        let e = errno::errno();
-        if let libc::EEXIST = e.0 {
-            return Err(StorageError::AlreadyExists(String::from(to)));
+    pub fn atomic_rename(from: &str, to: &str) -> Result<(), StorageError> {
+        // doing best effort in windows since there is no native atomic rename support
+        if std::fs::metadata(to).is_ok() {
+            return Err(StorageError::AlreadyExists(to.to_string()));
         }
+        std::fs::rename(from, to).map_err(|e| {
+            StorageError::other_std_io_err(format!("failed to rename {} to {}: {}", from, to, e))
+        })
+    }
+}
 
-        return Err(StorageError::Io {
-            source: custom_io_error(format!("{}", e)),
-        });
+#[cfg(unix)]
+mod imp {
+    use super::*;
+    use std::ffi::CString;
+
+    #[cfg(target_os = "linux")]
```

Comment:
Linux specific code in the unix-specific implementation? I'm quite confused by why these libc functions need to be mapped in? There's a high likelihood this could break on BSDs, non-glibc implementations.

(I haven't tested it of course, so you may know something I don't)

Isn't there a crate that abstracts this behavior which doesn't require the `unsafe` below?  E.g. https://docs.rs/crate/atomicwrites/0.3.0


Of course, if you've already gone down this path and this is the best possible option, then so be it, I would appreciate some commentary in the code explaining why "this must be"

Reply:
platform specific atomic rename needs to be implemented per platform unfortunately, so BSD and non-glibc users will just need to send PRs for the platforms they want to use, just like windows.

I looked into atomicwrites before, their implementation is not ideal. It requires 3 system calls to achieve what can be done in 1 system call, which is the implementation we have here. I also don't think their implementation is correct, because their implementation started with a hardlink function call, which based on the std rust doc, it doesn't error out if destination path already exists.

The unsafe code we have here is scoped to only system call invocation, which is basically what happens when you call `std::fs::rename`. So even though atomicwrites has less unsafe code on the surface, it's just hidden under the std lib implementation.

Reply:
@rtyler 
> why these libc functions need to be mapped in

Because they're missing in libc crate and PR for introducing them is still in open

Reply:
well https://github.com/rust-lang/libc/pull/2116#issuecomment-801511926 seems like they're merged this a couple of days ago

Reply:
ha nice, filed https://github.com/delta-io/delta-rs/issues/143 as follow up